### PR TITLE
Chore: Move voip's Wrap-up and On-hold functionality to EE (Backend)

### PR DESCRIPTION
**Pull request by murtaza98**

### CODE REVIEW

**Comment by KevLehman** [comment]
What if, instead of using callbacks, we create an enterprise service that overrides the close function on this service? Like with `overrideClassOnLicense` func, and then on that function we do the things to store the close room message 👀 

That way we don't need to use callbacks inside services

File: apps/meteor/server/services/omnichannel-voip/service.ts [hunk 1]
```diff
@@ -267,87 +268,75 @@ export class OmnichannelVoipService extends ServiceClassInternal implements IOmn
...
-	}
-
-	// Comment can be used to store wrapup call data
 	async closeRoom(
 		closerParam: ILivechatVisitor | ILivechatAgent,
 		room: IVoipRoom,
 		user: IUser,
-		comment?: string,
-		tags?: string[],
+		options: { comment?: string; tags?: string[] },
```
Comment (KevLehman) [inline apps/meteor/server/services/omnichannel-voip/service.ts:275]: Should options be optional? 👀

File: apps/meteor/server/services/omnichannel-voip/service.ts [hunk 2]
```diff
@@ -267,87 +268,75 @@ export class OmnichannelVoipService extends ServiceClassInternal implements IOmn
...
+		return true;
+	}
+
+	async getRoomClosingData(
+		closerParam: ILivechatVisitor | ILivechatAgent,
+		room: IVoipRoom,
+		options: { comment?: string; tags?: string[] },
+		sysMessageId: 'voip-call-wrapup' | 'voip-call-ended-unexpectedly',
+	): Promise<{ closeInfo: IRoomClosingInfo; closeSystemMsgData: any }> {
+		console.log('Calling community version');
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel-voip/service.ts:306]:
```suggestion
```

[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/server/services/omnichannel-voip/service.ts]: :see_no_evil:

File: apps/meteor/server/services/omnichannel-voip/service.ts (hunk 1 above)
Comment (murtaza98) [reply apps/meteor/server/services/omnichannel-voip/service.ts]: I see no issues with that, let me do that...

**Comment by KevLehman** [comment]
CI is failing 😬

**Review by KevLehman — COMMENTED** [review]
Currently on CE, it shows you the modal to close the conversation but doesn't store the data. Would be better to not to show the modal at all 👀 

The other thing: when looking at the calls panel, the call has both values as "Not available", would be good to not to show them if there's no valid license, maybe @MartinSchoeler can help you with how to achieve that 🤗

File: apps/meteor/app/api/server/v1/voip/rooms.ts [hunk 3]
```diff
@@ -218,10 +218,12 @@ API.v1.addRoute(
 			check(this.bodyParams, {
 				rid: String,
 				token: String,
-				comment: Match.Maybe(String),
-				tags: Match.Maybe([String]),
+				options: Match.Maybe({
+					comment: Match.Maybe(String),
+					tags: Match.Maybe([String]),
+				}),
```
Comment (ggazzo) [inline apps/meteor/app/api/server/v1/voip/rooms.ts:224]: what is the point of pass an object option instead shallow props?

File: packages/core-typings/src/IMessage/IMessage.ts [hunk 4]
```diff
@@ -232,3 +232,5 @@ export type IMessageInbox = IMessage & {
 
 export const isIMessageInbox = (message: IMessage): message is IMessageInbox => 'email' in message;
 export const isVoipMessage = (message: IMessage): message is IVoipMessage => 'voipData' in message;
+
+export type IOmniRoomClosingMessage = Pick<IMessage, 't' | 'groupable'> & Partial<Pick<IMessage, 'msg'>>;
```
Comment (ggazzo) [inline packages/core-typings/src/IMessage/IMessage.ts:236]: looks wrong, this looks more like a `system message` than IOminichannelRoomClosingSomething
I think this is a internal type, not a core one

**Review by ggazzo — CHANGES_REQUESTED** [review]
could you check my comment? thanks

**Comment by murtaza98** [comment]
> Currently on CE, it shows you the modal to close the conversation but doesn't store the data. Would be better to not to show the modal at all eyes
> 
> The other thing: when looking at the calls panel, the call has both values as "Not available", would be good to not to show them if there's no valid license, maybe @MartinSchoeler can help you with how to achieve that hugs

Well, both these behaviours were actually intentional since I was planning to keep this PR isolated to backend only change. However, I don't mind adding the frontend changes too in here if the frontend guys are okay with it. @MartinSchoeler WDYT?

File: apps/meteor/app/api/server/v1/voip/rooms.ts (hunk 3 above)
Comment (murtaza98) [reply apps/meteor/app/api/server/v1/voip/rooms.ts:224]: @ggazzo  I went with this after seeing a pattern across our code-base like [here](https://github.com/RocketChat/Rocket.Chat/blob/9578e15300465ff6f1074088b3b35366c567bb89/apps/meteor/app/livechat/server/methods/closeRoom.js#L8) where any params used by EE licence were bundled together. But let me know if you think we should still change this

File: packages/core-typings/src/IMessage/IMessage.ts (hunk 4 above)
Comment (murtaza98) [reply packages/core-typings/src/IMessage/IMessage.ts]: Changes done :heavy_check_mark:

File: apps/meteor/server/services/omnichannel-voip/helper.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import { IVoipRoom } from '@rocket.chat/core-typings';
+
+import { PbxEvent } from '../../../app/models/server/raw';
+
+export const calculateOnHoldTimeForRoom = async (room: IVoipRoom, closedAt: Date): Promise<number> => {
+	if (!room || !room.callUniqueId) {
```
Comment (ggazzo) [inline apps/meteor/server/services/omnichannel-voip/helper.ts:6]: room is not optional there is no need to test if its empty or not

**Comment by murtaza98** [comment]
> Currently on CE, it shows you the modal to close the conversation but doesn't store the data. Would be better to not to show the modal at all eyes
> 
> The other thing: when looking at the calls panel, the call has both values as "Not available", would be good to not to show them if there's no valid license, maybe @MartinSchoeler can help you with how to achieve that hugs

Hi @KevLehman The Frontend guys would like to keep this PR isolated from any frontend changes so it is easier to review. So what do you say about that? Should we merge this after this month's release?

**Comment by murtaza98** [comment]
moving this to draft since some changes are also required for On-hold feature migration

**Comment by murtaza98** [comment]
> moving this to draft since some changes are also required for On-hold feature migration

Done with this change. This PR is again ready for review :smile:

**Comment by KevLehman** [comment]
Noticed you moved some files to `livechat-enterprise`, could you please create a `voip-enterprise` instead? 🙈 since the functionality you moved was not from livechat but from voip 👀

**Comment by murtaza98** [comment]
> Noticed you moved some files to `livechat-enterprise`, could you please create a `voip-enterprise` instead? see_no_evil since the functionality you moved was not from livechat but from voip eyes

That's a good idea. I've moved them into their own folder now :+1: 

PS: Seems like we're facing the same build issue on this PR too :(

**Review by KevLehman — COMMENTED** [review]
Apart from that comment, function-wise this seems to be working as expected 👀

**Review by KevLehman — APPROVED** [review]

**Review by KevLehman — DISMISSED** [review]

**Review by ggazzo — APPROVED** [review]

**Review by KevLehman — APPROVED** [review]

**Pull request merged**